pkg/core/native/ledger.go

```diff
@@ -0,0 +1,230 @@
+}
+
+// currentHash implements currentHash SC method.
+func (l *Ledger) currentHash(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	return stackitem.Make(ic.Chain.CurrentBlockHash().BytesBE())
+}
+
+// currentIndex implements currentIndex SC method.
+func (l *Ledger) currentIndex(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	return stackitem.Make(ic.Chain.BlockHeight())
+}
+
+// getBlock implements getBlock SC method.
+func (l *Ledger) getBlock(ic *interop.Context, params []stackitem.Item) stackitem.Item {
+	hash := getBlockHashFromItem(ic.Chain, params[0])
+	block, err := ic.Chain.GetBlock(hash)
+	if err != nil || !isTraceableBlock(ic.Chain, block.Index) {
+		return stackitem.Null{}
+	}
+	return BlockToStackItem(block)
+}
+
+// getTransaction returns transaction to the SC.
+func (l *Ledger) getTransaction(ic *interop.Context, params []stackitem.Item) stackitem.Item {
+	tx, h, err := getTransactionAndHeight(ic.DAO, params[0])
+	if err != nil || !isTraceableBlock(ic.Chain, h) {
+		return stackitem.Null{}
+	}
+	return TransactionToStackItem(tx)
+}
+
+// getTransactionHeight returns transaction height to the SC.
+func (l *Ledger) getTransactionHeight(ic *interop.Context, params []stackitem.Item) stackitem.Item {
+	_, h, err := getTransactionAndHeight(ic.DAO, params[0])
+	if err != nil || !isTraceableBlock(ic.Chain, h) {
+		return stackitem.Make(-1)
+	}
+	return stackitem.Make(h)
+}
+
+// getTransactionFromBlock returns transaction with the given index from the
+// block with height or hash specified.
+func (l *Ledger) getTransactionFromBlock(ic *interop.Context, params []stackitem.Item) stackitem.Item {
+	hash := getBlockHashFromItem(ic.Chain, params[0])
+	index, err := params[1].TryInteger()
+	if err != nil {
+		panic(err)
+	}
+	block, err := ic.Chain.GetBlock(hash)
+	if err != nil || !isTraceableBlock(ic.Chain, block.Index) {
+		return stackitem.Null{}
+	}
+	indexi := index.Int64()
+	if !index.IsInt64() || indexi < 0 || indexi >= int64(len(block.Transactions)) {
+		panic("wrong transaction index")
+	}
+	return TransactionToStackItem(block.Transactions[indexi])
+}
+
+// isTraceableBlock defines whether we're able to give information about
+// the block with index specified.
+func isTraceableBlock(bc blockchainer.Blockchainer, index uint32) bool {
+	height := bc.BlockHeight()
+	MaxTraceableBlocks := bc.GetConfig().MaxTraceableBlocks
+	return index <= height && index+MaxTraceableBlocks > height
+}
+
+// getBlockHashFromItem converts given stackitem.Item to block hash using given
+// Blockchainer if needed. Interop functions accept both block numbers and
+// block hashes as parameters, thus this function is needed. It's supposed to
+// be called within VM context, so it panics if anything goes wrong.
+func getBlockHashFromItem(bc blockchainer.Blockchainer, item stackitem.Item) util.Uint256 {
+	bigindex, err := item.TryInteger()
+	if err == nil && bigindex.IsInt64() {
```

Comment:
`IsUint64` and don't check for `<0` ?

Reply:
`toUint32` fits precisely here, and yes, these helpers make code a bit simpler.

Reply:
Except we can't panic here, so I'll leave this one as is.